### PR TITLE
Multiplayer Resimulation for Client Fired Linear Weapons (Rollback)

### PR DESCRIPTION
**Pull request by JohnAFernandez**
This creates a packet that sends the relative position and orientation info for a primary or dumbfire missile shot.  The client creates this packet when firing those two types of weapons and sends it to the server.  The server goes back to the frame that the client fired the weapon and quickly re-simulates all the frames in-between.

This includes collision detection, restoring positions, orientations and velocities, but not subsystem orientations.

Now that the object update packet is merged, note that this is mainly to ensure that high ping clients can fire primary shots with the same accuracy on server and client.

### CODE REVIEW

**Comment by JohnAFernandez** [comment]
If you started a review, I simplified one of the functions, by sending an int instead of a ubyte in the new packet type, so you may want to take another look.

File: code/network/multimsgs.cpp
Comment (notimaginative) [inline code/network/multimsgs.cpp:7635]: What happens if this wraps? Given sufficient lag it would be possible so if this would cause a problem then I'd suggest making this a ushort.

File: code/object/objcollide.cpp
Comment (notimaginative) [inline code/object/objcollide.cpp:1026]: This should be `&Collision_list`. Otherwise you're just copying the passed list to this one and sorting the copy. The list you intended to sort remains unchanged.

Comment (JohnAFernandez) [reply code/object/objcollide.cpp:1026]: Well that's inefficient!

**Comment by JohnAFernandez** [comment]
I'll get started on these changes because it turns out the fotg bug had nothing to do with this pr whatsoever.  Autoaim in multi needs to be fixed.

File: code/network/multimsgs.cpp
Comment (JohnAFernandez) [reply code/network/multimsgs.cpp:7635]: Changes made....

File: code/network/multi_obj.cpp
Comment (JohnAFernandez) [reply code/network/multi_obj.cpp:601]: Changes made.  Need to double check that I didn't break anything with these changes.

File: code/network/multi_obj.cpp
Comment (JohnAFernandez) [reply code/network/multi_obj.cpp:592]: Changes made.

File: code/object/objcollide.cpp
Comment (JohnAFernandez) [reply code/object/objcollide.cpp:1026]: Changes made...

**Comment by JohnAFernandez** [comment]
Now to retest....

**Comment by JohnAFernandez** [comment]
Testing went fine!  I made a couple of edits that are different from before @notimaginative

File: code/object/objcollide.cpp
Comment (notimaginative) [inline code/object/objcollide.cpp:1026]: You might need a `nullptr` check on this. It really shouldn't be needed the way it's used, but you never know. I'll leave it up to you.

But if you do add it I might suggest making the default be `nullptr` rather than pointing to `Collision_sort_list`, then if it's `nullptr` assign `Collision_sort_list` at that time. That makes it a little easier to understand what's going to happen without having to look at another part of the code. That's more of a personal preference/style though.

**Review by notimaginative — APPROVED** [review]
LGTM, as far as I can tell. Maybe someone else will give it a look and notice something that I didn't.

File: code/object/objcollide.cpp
Comment (JohnAFernandez) [reply code/object/objcollide.cpp:1026]: I'm all for more readability, so I'll make that change.

**Comment by JohnAFernandez** [comment]
I'll give it a day before I merge.  Speak now or forever hold your peace!

**Pull request merged**
